File: src/main/java/toby/command/commands/music/LeaveCommand.java
package toby.command.commands.music;

import net.dv8tion.jda.api.Permission;
import net.dv8tion.jda.api.entities.Guild;
import net.dv8tion.jda.api.entities.GuildVoiceState;
import net.dv8tion.jda.api.entities.Member;
import net.dv8tion.jda.api.entities.channel.unions.AudioChannelUnion;
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent;
import net.dv8tion.jda.api.managers.AudioManager;
import toby.command.CommandContext;
import toby.jpa.dto.ConfigDto;
import toby.jpa.dto.UserDto;
import toby.jpa.service.IConfigService;
import toby.lavaplayer.GuildMusicManager;
import toby.lavaplayer.PlayerManager;

import static toby.command.ICommand.invokeDeleteOnMessageResponse;

public class LeaveCommand implements IMusicCommand {
    private final IConfigService configService;

    public LeaveCommand(IConfigService configService) {
        this.configService = configService;
    }

    @Override
    public void handle(CommandContext ctx, UserDto requestingUserDto, Integer deleteDelay) {
        handleMusicCommand(ctx, PlayerManager.getInstance(), requestingUserDto, deleteDelay);
    }

    @Override
    public void handleMusicCommand(CommandContext ctx, PlayerManager instance, UserDto requestingUserDto, Integer deleteDelay) {
        final SlashCommandInteractionEvent event = ctx.getEvent();
        event.deferReply().queue();
        final Member self = ctx.getSelfMember();
        final GuildVoiceState selfVoiceState = self.getVoiceState();

        if (!requestingUserDto.hasMusicPermission()) {
            sendErrorMessage(event, deleteDelay);
            return;
        }

        if (isInvalidChannelStateForCommand(deleteDelay, event, selfVoiceState)) return;

        final Member member = ctx.getMember();
        final GuildVoiceState memberVoiceState = member.getVoiceState();


        Guild guild = event.getGuild();
        final AudioManager audioManager = guild.getAudioManager();
        AudioChannelUnion memberChannel = memberVoiceState.getChannel();
        GuildMusicManager musicManager = instance.getMusicManager(guild);

        if (PlayerManager.getInstance().isCurrentlyStoppable() || member.hasPermission(Permission.KICK_MEMBERS)) {
            String volumePropertyName = ConfigDto.Configurations.VOLUME.getConfigValue();
            ConfigDto databaseConfig = configService.getConfigByName(volumePropertyName, event.getGuild().getId());
            int defaultVolume = databaseConfig!=null ? Integer.parseInt(databaseConfig.getValue()) : 100;
            musicManager.getScheduler().setLooping(false);
            musicManager.getScheduler().getQueue().clear();
            musicManager.getAudioPlayer().stopTrack();
            musicManager.getAudioPlayer().setVolume(defaultVolume);
            audioManager.closeAudioConnection();
            event.getHook().sendMessageFormat("Disconnecting from `\uD83D\uDD0A %s`", memberChannel.getName()).queue(invokeDeleteOnMessageResponse(deleteDelay));
        } else {
            IMusicCommand.sendDeniedStoppableMessage(event, musicManager, deleteDelay);
        }
    }


    private static boolean isInvalidChannelStateForCommand(Integer deleteDelay, SlashCommandInteractionEvent event, GuildVoiceState selfVoiceState) {
        if (!selfVoiceState.inAudioChannel()) {
            event.getHook().sendMessage("I'm not in a voice channel, somebody shoot this guy").queue(invokeDeleteOnMessageResponse(deleteDelay));
            return true;
        }
        return false;
    }


    @Override
    public String getName() {
        return "leave";
    }

    @Override
    public String getDescription() {
        return "Makes the TobyBot leave the voice channel it's currently in";
    }

}

File: src/main/kotlin/toby/command/commands/music/LeaveCommand.kt
package toby.command.commands.music

import net.dv8tion.jda.api.entities.GuildVoiceState
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent
import toby.command.CommandContext
import toby.command.ICommand.Companion.invokeDeleteOnMessageResponse
import toby.jpa.dto.ConfigDto
import toby.jpa.dto.UserDto
import toby.jpa.service.IConfigService
import toby.lavaplayer.PlayerManager

class LeaveCommand(private val configService: IConfigService) : IMusicCommand {

    override fun handle(ctx: CommandContext, requestingUserDto: UserDto, deleteDelay: Int?) {
        handleMusicCommand(ctx, PlayerManager.instance, requestingUserDto, deleteDelay)
    }

    override fun handleMusicCommand(
        ctx: CommandContext,
        instance: PlayerManager,
        requestingUserDto: UserDto,
        deleteDelay: Int?
    ) {
        val event = ctx.event
        event.deferReply().queue()

        if (!requestingUserDto.musicPermission) {
            sendErrorMessage(event, deleteDelay!!)
            return
        }

        val selfVoiceState = ctx.selfMember?.voiceState
        if (isInvalidChannelStateForCommand(deleteDelay, event, selfVoiceState)) return

        val guild = event.guild!!
        val audioManager = guild.audioManager
        val musicManager = instance.getMusicManager(guild)

        if (instance.isCurrentlyStoppable || requestingUserDto.superUser) {
            handleStopCommand(event, selfVoiceState, guild, musicManager, audioManager, deleteDelay)
        } else {
            IMusicCommand.sendDeniedStoppableMessage(event.hook, musicManager, deleteDelay)
        }
    }

    private fun handleStopCommand(
        event: SlashCommandInteractionEvent,
        selfVoiceState: GuildVoiceState?,
        guild: net.dv8tion.jda.api.entities.Guild,
        musicManager: toby.lavaplayer.GuildMusicManager,
        audioManager: net.dv8tion.jda.api.managers.AudioManager,
        deleteDelay: Int?
    ) {
        val volumePropertyName = ConfigDto.Configurations.VOLUME.configValue
        val defaultVolume = configService.getConfigByName(volumePropertyName, guild.id)?.value?.toInt() ?: 100

        musicManager.scheduler.apply {
            isLooping = false
            queue.clear()
        }
        musicManager.audioPlayer.apply {
            stopTrack()
            volume = defaultVolume
        }
        audioManager.closeAudioConnection()

        event.hook
            .sendMessage("Disconnecting from `\uD83D\uDD0A ${selfVoiceState?.channel?.name}`")
            .queue(invokeDeleteOnMessageResponse(deleteDelay ?: 0))
    }

    override val name: String
        get() = "leave"

    override val description: String
        get() = "Makes the TobyBot leave the voice channel it's currently in"

    companion object {
        private fun isInvalidChannelStateForCommand(
            deleteDelay: Int?,
            event: SlashCommandInteractionEvent,
            selfVoiceState: GuildVoiceState?
        ): Boolean {
            return if (!selfVoiceState?.inAudioChannel()!!) {
                event.hook.sendMessage("I'm not in a voice channel, somebody shoot this guy")
                    .queue(invokeDeleteOnMessageResponse(deleteDelay!!))
                true
            } else {
                false
            }
        }
    }
}